api/samples/memval_simple.c

```diff
@@ -280,6 +279,10 @@ instrument_post_write(void *drcontext, instrlist_t *ilist, instr_t *where, opnd_
     drx_buf_insert_buf_memcpy(drcontext, write_buffer, ilist, where, reg_ptr, reg_addr,
                               stride);
 
+    drx_buf_insert_load_buf_ptr(drcontext, trace_buffer, ilist, where, reg_ptr);
```

Comment:
The drx_buf_insert_buf_memcpy docs say it updates the buffer pointer -- does it clobber reg_ptr too and that's why it has to be reloaded?  But I don't understand: updating the pointer for the address, which was written pre-app-instr, would have to be done *before* storing the value here: else the value will clobber the address, and then this extra advance will just have uninitialized data in the buffer.

Reply:
There are two different buffers here -- `trace_buffer` that stores addr, size, type of the write, and `write_buffer` that stores the actual written value. The former is written before the app write by `instrument_mem`, whereas the latter is written after the app write by `instrument_post_write`.

> does it clobber reg_ptr too and that's why it has to be reloaded?

Before L282, `reg_ptr` has the `write_buffer` ptr, but we need the `trace_buffer` ptr, so we reload.

> else the value will clobber the address, and then this extra advance will just have uninitialized data in the buffer.

The written value is stored in a different buffer, so it won't clobber the address.

Reply:
By delaying the advancement of the `trace_buffer` ptr, in a way we delay committing the <addr, type, size> written by `instrument_mem`. So if the app write fails for some reason and the app recovers from the error (like `drx-scattergather.c` does), there'll be no mismatch between `trace_buffer` and `write_buffer`, and the DR client will be able to continue on.

Reply:
This adds an extra load for a very rare path: most app stores do not fault.  If the buffer pointer discrepancy can be handled in the fault handler, that seems a superior solution, rather than forcing every single non-faulting store to pay extra overhead, right?

Reply:
I agree. I considered handling this case in a fault handler, but decided against the extra complexity in this sample client. If performance is a bigger concern here, sure I can use the fault handler strategy.

Reply:
How much more complex is it?  It looks like this client already has a fault handler.  Oh wait -- that's the trace buffer fault, not the app fault.  OK, I was thinking it would just be one line in the existing handler, but it would take a whole new event.  Yeah, doing it this way with a comment that this is extra common path overhead that could be avoided with an app fault handler seems fine.

Reply:
Thanks. Done.